From jason.wessel@windriver.com  Thu Jul  2 16:00:37 2009
From: Jason Wessel <jason.wessel@windriver.com>
Date: Mon, 22 Jun 2009 11:32:21 -0500
Subject: USB: serial: regression fix to move sysrq from hot path
To: greg@kroah.com
Cc: Jason Wessel <jason.wessel@windriver.com>
Message-ID: <1245688342-26373-3-git-send-email-jason.wessel@windriver.com>


A performance regression was introduced by commit:
98fcb5f78165b8a3d93870ad7afd4d9ebbb8b43a

The sysrq handling should only get executed if the attached usb serial
device is acting as a serial system console.  A serial system console
has a very low input throughput vs a 3g usb modem which pushes bytes
through the same interface at a high rate.

Entire strings of output can be processed via the tty input when the
device is not a console.

Signed-off-by: Jason Wessel <jason.wessel@windriver.com>
Acked-by: Alan Cox <alan@linux.intel.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/usb/serial/generic.c |   16 ++++++++++++----
 1 file changed, 12 insertions(+), 4 deletions(-)

--- a/drivers/usb/serial/generic.c
+++ b/drivers/usb/serial/generic.c
@@ -425,11 +425,19 @@ static void flush_and_resubmit_read_urb(
 		goto done;
 
 	/* Push data to tty */
-	for (i = 0; i < urb->actual_length; i++, ch++) {
-		if (!usb_serial_handle_sysrq_char(port, *ch))
-			tty_insert_flip_char(tty, *ch, TTY_NORMAL);
+	if (unlikely(port->console)) {
+		for (i = 0; i < urb->actual_length; i++, ch++) {
+			if (!usb_serial_handle_sysrq_char(port, *ch))
+				tty_insert_flip_char(tty, *ch, TTY_NORMAL);
+		}
+		tty_flip_buffer_push(tty);
+	} else if (urb->actual_length) {
+		i = tty_buffer_request_room(tty, urb->actual_length);
+		if (i) {
+			tty_insert_flip_string(tty, urb->transfer_buffer, i);
+			tty_flip_buffer_push(tty);
+		}
 	}
-	tty_flip_buffer_push(tty);
 	tty_kref_put(tty);
 done:
 	usb_serial_generic_resubmit_read_urb(port, GFP_ATOMIC);
